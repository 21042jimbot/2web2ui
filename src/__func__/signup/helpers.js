import MockRecaptcha from 'src/__testHelpers__/mockRecaptcha';
import mountRoute from 'src/__testHelpers__/mountRoute';
import getFormFiller from 'src/__testHelpers__/fill';
import config from 'src/config';

jest.mock('react-recaptcha', () => MockRecaptcha);

export const freePlanPredicate = plan => /free/.test(plan.code);

class SignupFlow {
  constructor() {
    config.featureFlags.has_signup = true;
  }

  async _mount() {
    this.page = await mountRoute('/join', { authenticated: false });
    this.formFiller = getFormFiller(this.page.wrapper);
    for (const prop in this.page) {
      this[prop] = this.page[prop];
    }
  }

  async signup({ optIn = false } = {}) {
    this.formFiller([
      { name: 'first_name', value: 'Firsty' },
      { name: 'last_name', value: 'Ferret' },
      { name: 'email', value: 'test-username@example.com' },
      { name: 'password', value: 'test-password' },
      { name: 'tou_accepted', value: true, type: 'checkbox' },
      { name: 'email_opt_in', value: optIn, type: 'checkbox' },
    ]);
    await this.page.simulate('button[id="submit"]', 'click');
  }
}

export default async () => {
  const flow = new SignupFlow();
  await flow._mount();
  return flow;
};
