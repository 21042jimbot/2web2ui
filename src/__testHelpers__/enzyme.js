export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
