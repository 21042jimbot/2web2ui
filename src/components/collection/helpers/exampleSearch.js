import _ from 'lodash';
const MAX_EXAMPLE_VALUE_LENGTH = 50;

export const isValidExampleValue = (row, key) => (
  typeof row[key] !== 'undefined' &&
  row[key] !== null &&
  row[key] !== ''
);

/**
 * Takes a data set and a list of example modifiers and provides real example
 * values from the given data set, returning an array of key:value combinations
 *
 * Args are (rows, options) so that memoization can be done easily on the rows list
 *
 * @param {Array} rows Data set to use for deriving example values
 * @param {Object} options
 * @param {Array|String} options.exampleModifiers List of keys to use as example modifiers, or hard-coded example modifier string
 * @param {Array} options.keyMap Hash for mapping friendly keys to their real key names in the data set
 */
export function createExampleSearch(rows, { exampleModifiers = [], keyMap = {}, MAX_LENGTH = MAX_EXAMPLE_VALUE_LENGTH }) {
  // Allows passing in of a string version of thing:value, for complicated cases
  if (typeof exampleModifiers === 'string') {
    return exampleModifiers;
  }

  // Derive examples from data set
  const converted = exampleModifiers.map((modifier) => {
    const realKey = keyMap[modifier] || modifier;
    const rowWithValue = rows.find((row) => isValidExampleValue(row, realKey));

    // If no valid example values were found for this modifier key, skip it (we'll filter out false values later)
    if (!rowWithValue) {
      return false;
    }

    let value = rowWithValue[realKey];

    // If the value is a string longer than our max, truncate and trim trailing spaces
    if (typeof value === 'string' && value.length > MAX_LENGTH) {
      value = value.substr(0, MAX_LENGTH).trim();
    }

    // If the stringified value contains spaces, surround value in double quotes
    if (String(value).includes(' ')) {
      value = `"${value}"`;
    }

    // Return the key:value pair in string format
    return `${modifier}:${value}`;
  });

  // Filter out falsey values
  const results = converted.filter((pair) => pair);

  // Don't guess at keys that are searchable or return an empty list, use a generic example instead
  if (results.length === 0) {
    return ['myKey:123dF9', 'myKey:"Some Value"', 'myKey:true'];
  }

  return results;
}

// Memoizes createExampleSearch on first argument, "rows", to prevent re-calculation
export const memoizedCreateExampleSearch = _.memoize(createExampleSearch);

// Get a random search pair, e.g. key:value
export const getRandomExampleSearch = ({ rows, ...options }) => _.sample(memoizedCreateExampleSearch(rows, options));
