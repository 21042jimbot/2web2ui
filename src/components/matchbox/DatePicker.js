import React from 'react';
import { useHibana } from 'src/context/HibanaContext';
import { DatePicker as DatePickerComponent } from '@sparkpost/matchbox-hibana';

function DatePicker(props) {
  const [{ isHibanaEnabled }] = useHibana();
  if (!isHibanaEnabled) {
    throw new Error(
      'DatePicker component not available in original matchbox. Please remove or restrict to Hibana',
    );
  }
  return <DatePickerComponent {...props} />;
}

export default DatePicker;
