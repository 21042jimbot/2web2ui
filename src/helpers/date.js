import moment from 'moment';
import config from 'src/config';


export const relativeDateOptions = [
  { value: 'hour', label: 'Last Hour' },
  { value: 'day', label: 'Last 24 Hours' },
  { value: '7days', label: 'Last 7 Days' },
  { value: '30days', label: 'Last 30 Days' },
  { value: '90days', label: 'Last 90 Days' },
  { value: 'custom', label: 'Custom' }
];

export const relativeDateOptionsIndexed = relativeDateOptions.reduce((result, { value, label }) => {
  result[value] = label;
  return result;
}, {});

/**
 * Takes a date string and returns the end of that day (11:59PM)
 *
 * If preventFuture is true and the given day IS the current day,
 * it returns the current time, i.e. the closest to the end
 * of the day without going into the future.
 *
 * @param {String} date - date string to base date on
 * @return {Date}
 */
export function getEndOfDay(date, { preventFuture } = {}) {
  const now = moment();
  const end = new Date(date);

  if (preventFuture && now.diff(end, 'days') === 0) {
    return now.toDate();
  }

  end.setHours(23, 59, 59, 999);
  return end;
}

export function getStartOfDay(date) {
  const start = new Date(date);
  start.setHours(0, 0, 0, 0);
  return start;
}

export function isSameDate(a, b) {
  return (a instanceof Date) && (b instanceof Date) && (a.getTime() === b.getTime());
}

export function getRelativeDates(range) {
  const now = moment.utc();
  const to = now.toDate();

  switch (range) {
    case 'hour':
      return { to, from: moment(to).subtract(1, 'hour').toDate(), relativeRange: range };

    case 'day':
      return { to, from: moment(to).subtract(1, 'day').toDate(), relativeRange: range };

    case '7days':
      return { to, from: moment(to).subtract(7, 'day').toDate(), relativeRange: range };

    case '30days':
      return { to, from: moment(to).subtract(30, 'day').toDate(), relativeRange: range };

    case '90days':
      return { to, from: moment(to).subtract(90, 'day').toDate(), relativeRange: range };

    case 'custom':
      return { relativeRange: range };

    default:
      return {};
  }
}

export function formatDate(date) {
  return moment(date).format(config.dateFormat);
}

export function formatTime(time) {
  return moment(time).format(config.timeFormat);
}

export function formatDateTime(datetime) {
  return `${formatDate(datetime)}, ${formatTime(datetime)}`;
}
