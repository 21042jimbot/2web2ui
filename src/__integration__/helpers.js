import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'src/store';
import { Provider } from 'react-redux';
import _ from 'lodash';

export const asyncFlush = () => new Promise((resolve) => setImmediate(resolve));

export const login = (store) => store.dispatch({
  type: 'LOGIN_SUCCESS',
  payload: {
    access_token: 'very-real-access-token'
  }
});

export async function setupForm(tree) {
  const store = configureStore();
  login(store);

  const mounted = mount(
    <Provider store={store}>
      <MemoryRouter>
        {tree}
      </MemoryRouter>
    </Provider>
  );

  const change = (index, value) => {
    const el = mounted.find('Field').at(index);
    let control = el.find('input');
    if (control.length === 0) {
      control = el.find('select');
    }
    control.simulate('change', { target: { value }});
    mounted.update();
  };

  await asyncFlush();
  mounted.update();

  return {
    mounted,
    store,
    change,
    asyncFlush,
    fillInOrder: (fields) => fields.map((val, i) => change(i, val)),
    submit: async () => {
      mounted.find('form').simulate('submit');
      return asyncFlush();
    },
    /* eslint-disable-next-line no-console */
    debug: (path = 'form') => console.log(JSON.stringify(_.get(store.getState(), path), null, 2))
  };
}
