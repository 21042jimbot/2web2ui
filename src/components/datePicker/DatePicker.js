/* eslint max-lines: ["error", 200] */
import React, { Component } from 'react';
import { subMonths, format } from 'date-fns';
import { getStartOfDay, getEndOfDay, getRelativeDateOptions } from 'src/helpers/date';
import { roundBoundaries } from 'src/helpers/metrics';
import { Button, TextField, Select, Popover, WindowEvent } from '@sparkpost/matchbox';
import DateSelector from 'src/components/dateSelector/DateSelector';
import ManualEntryForm from './ManualEntryForm';
import { FORMATS } from 'src/constants';
import styles from './DatePicker.module.scss';
import PropTypes from 'prop-types';

export default class AppDatePicker extends Component {

  DATE_FORMAT = FORMATS.LONG_DATETIME;
  state = {
    showDatePicker: false,
    selecting: false,
    selected: { }
  }

  componentDidMount() {
    this.syncTimeToState(this.props);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.from !== this.props.from || nextProps.to !== this.props.to) {
      this.syncTimeToState(nextProps);
    }
  }

  // Sets local state from reportOptions redux state - need to separate to handle pre-apply state
  syncTimeToState = ({ from, to }) => {
    this.setState({ selected: { from, to }});
  }

  // Closes popover on escape, submits on enter
  handleKeyDown = (e) => {
    if (!this.state.showDatePicker) {
      return;
    }

    if (e.key === 'Escape') {
      this.cancelDatePicker();
    }

    if (!this.state.selecting && e.key === 'Enter') {
      this.handleSubmit();
    }
  }

  handleDayKeyDown = (day, modifiers, e) => {
    this.handleKeyDown(e);
    e.stopPropagation();
  }

  cancelDatePicker = () => {
    this.syncTimeToState(this.props);
    this.setState({ showDatePicker: false });
  }

  showDatePicker = () => {
    this.setState({ showDatePicker: true });
  }

  handleDayClick = (clicked) => {
    const { selecting, selected } = this.state;
    const dates = selecting
      ? selected
      : { from: getStartOfDay(clicked), to: getEndOfDay(clicked, { preventFuture: true }) };

    this.setState({
      selected: dates,
      beforeSelected: dates,
      selecting: !selecting
    });
  }

  handleDayHover = (hovered) => {
    const { selecting } = this.state;

    if (selecting) {
      this.setState({ selected: { ...this.getOrderedRange(hovered) }});
    }
  }

  getOrderedRange(newDate) {
    let { from, to } = this.state.beforeSelected;

    (from.getTime() <= newDate.getTime())
      ? to = getEndOfDay(newDate, { preventFuture: true })
      : from = getStartOfDay(newDate);

    const rounded = roundBoundaries(from, to);
    return { from: rounded.from.toDate(), to: rounded.to.toDate() };
  }

  handleSelectRange = (e) => {
    const value = e.currentTarget.value;

    if (value === 'custom') {
      this.setState({ showDatePicker: true });
    } else {
      this.setState({ showDatePicker: false });
      this.props.onChange({ relativeRange: value });
    }
  }

  handleFormDates = ({ from, to }, callback) => {
    this.setState({ selected: { from, to }}, () => callback());
  }

  handleSubmit = () => {
    this.setState({ showDatePicker: false, selecting: false });
    this.props.onChange({ ...this.state.selected, relativeRange: 'custom' });
  }

  render() {
    const { selected: { from, to }, showDatePicker } = this.state;
    const selectedRange = showDatePicker ? 'custom' : this.props.relativeRange;

    // allow for prop-level override of "now" (DI, etc.)
    const {
      now = new Date(),
      relativeDateOptions = [],
      disabled,
      datePickerProps = {},
      textFieldProps = {},
      dateFieldFormat,
      showPrecision
      showPresets = true,
      left
    } = this.props;
    const dateFormat = dateFieldFormat || this.DATE_FORMAT;

    const rangeSelect = showPresets
      ? <Select
        options={getRelativeDateOptions(relativeDateOptions)}
        onChange={this.handleSelectRange}
        value={selectedRange}
        disabled={disabled} />
      : null;

    const dateField = <TextField
      onClick={this.showDatePicker}
      connectLeft={rangeSelect}
      value={`${format(from, dateFormat)} â€“ ${format(to, dateFormat)}`}
      readOnly
      disabled={disabled}
      {...textFieldProps} />;

    return (
      <Popover
        wrapper='div'
        className={styles.Popover}
        trigger={dateField}
        onClose={this.cancelDatePicker}
        open={this.state.showDatePicker}
        left={left} >

        <DateSelector
          numberOfMonths={2}
          fixedWeeks
          enableOutsideDays={false}
          initialMonth={subMonths(now, 1)}
          toMonth={now}
          disabledDays={{ after: now }}
          onDayClick={this.handleDayClick}
          onDayMouseEnter={this.handleDayHover}
          onDayFocus={this.handleDayHover}
          onKeyDown={this.handleKeyDown}
          onDayKeyDown={this.handleDayKeyDown}
          selectedDays={this.state.selected}
          {...datePickerProps}
        />

        <ManualEntryForm selectDates={this.handleFormDates} onEnter={this.handleKeyDown} to={to} from={from} showPrecision={showPrecision}/>
        <Button primary onClick={this.handleSubmit} className={styles.Apply}>Apply</Button>
        <Button onClick={this.cancelDatePicker}>Cancel</Button>
        <WindowEvent event='keydown' handler={this.handleKeyDown} />
      </Popover>
    );
  }
}

AppDatePicker.propTypes = {
  now: PropTypes.instanceOf(Date),
  from: PropTypes.instanceOf(Date),
  to: PropTypes.instanceOf(Date),
  relativeRange: PropTypes.string,
  relativeDateOptions: PropTypes.arrayOf(PropTypes.string),
  showPrecision: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  datePickerProps: PropTypes.object,
  dateFieldFormat: PropTypes.string,
  disabled: PropTypes.bool,
  showPresets: PropTypes.bool
};
