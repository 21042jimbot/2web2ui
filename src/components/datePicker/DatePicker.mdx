# DatePicker

A text field paired with a `<select>` that allows users to select date ranges.

## Usage

```js
import React from 'react';
import DatePicker from 'src/components/datePicker/DatePicker';

const MyComponent = () => (
  <DatePicker
    relativeDateOptions={['hour', 'day', '7days', '30days', '90days', 'custom']}
    relativeRange="hour"
    from={new Date()}
    to={new Date()}
    disabled={false}
    onChange={() => console.log('change!')}
    roundToPrecision={true}
  />
);
```
