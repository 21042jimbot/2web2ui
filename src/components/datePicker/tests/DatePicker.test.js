import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';
import * as dateHelpers from 'src/helpers/date';
import DatePicker from '../DatePicker';
import utc from 'src/__testHelpers__/time';

jest.mock('src/helpers/date');
jest.mock('react-dom');
jest.mock('date-fns');

describe('Component: DatePicker', () => {

  let wrapper;
  let instance;
  let props;
  let mockNow;
  let mockFrom;

  beforeEach(() => {
    mockNow = utc({ year: 2017, month: 1, day: 15, hour: 12 });
    mockFrom = utc({ year: 2017, month: 1, day: 15, hour: 11 });
    props = {
      from: mockFrom,
      to: mockNow,
      relativeRange: 'day',
      relativeDateOptions: [],
      onChange: jest.fn(),
      now: mockNow,
      disabled: false
    };

    dateHelpers.getStartOfDay = jest.fn(() => 'start-of-day');
    dateHelpers.getEndOfDay = jest.fn(() => 'end-of-day');
    dateHelpers.getRelativeDateOptions = jest.fn(() => [1, 2, 3]);

    wrapper = shallow(<DatePicker {...props} />);
    instance = wrapper.instance();

    // spy on all instance methods
    _.functions(instance).forEach((f) => jest.spyOn(instance, f));
  });

  it('should render ok by default', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should properly mount', () => {
    jest.spyOn(window, 'addEventListener');

    instance.componentDidMount();
    expect(instance.syncTimeToState).toHaveBeenCalledTimes(1);
  });

  describe('syncTimeToState', () => {

    const before = { from: 'unchanged', to: 'unchanged' };
    const after = { from: new Date(), to: new Date() };

    it('should sync the state when from or to is changed', () => {
      wrapper.setState({ selected: before });
      expect(wrapper.state('selected')).toEqual(before);
      wrapper.setProps(after);
      expect(instance.syncTimeToState).toHaveBeenCalledTimes(1);
      expect(wrapper.state('selected')).toEqual(after);
    });

    it('shouldn\'t sync the state when from and to aren\'t changed', () => {
      wrapper.setState({ selected: before });
      wrapper.setProps({ other: 'stuff' });
      expect(instance.syncTimeToState).not.toHaveBeenCalled();
      expect(wrapper.state('selected')).toEqual(before);
    });

  });

  describe('handleKeyDown', () => {

    const ESCAPE = { key: 'Escape' };
    const ENTER = { key: 'Enter' };

    it('should do nothing if picker is closed already', () => {
      wrapper.setState({ showDatePicker: false });
      instance.handleKeyDown(ESCAPE);
      expect(instance.cancelDatePicker).not.toHaveBeenCalled();
      expect(instance.handleSubmit).not.toHaveBeenCalled();
      expect(wrapper.state('showDatePicker')).toEqual(false);
    });

    it('should cancel date picker on escape keydown', () => {
      wrapper.setState({ showDatePicker: true });
      instance.handleKeyDown(ESCAPE);
      expect(instance.cancelDatePicker).toHaveBeenCalledTimes(1);
      expect(instance.handleSubmit).not.toHaveBeenCalled();
      expect(wrapper.state('showDatePicker')).toEqual(false);
    });

    it('should submit on enter', () => {
      wrapper.setState({ showDatePicker: true, selecting: false });
      instance.handleKeyDown(ENTER);
      expect(instance.cancelDatePicker).not.toHaveBeenCalled();
      expect(instance.handleSubmit).toHaveBeenCalledTimes(1);
      expect(wrapper.state('showDatePicker')).toEqual(false);
    });

    it('should not submit if selecting is true', () => {
      wrapper.setState({ showDatePicker: true, selecting: true });
      instance.handleKeyDown(ENTER);
      expect(instance.cancelDatePicker).not.toHaveBeenCalled();
      expect(instance.handleSubmit).not.toHaveBeenCalled();
      expect(wrapper.state('showDatePicker')).toEqual(true);
    });

  });

  describe('handleDayKeyDown', () => {

    it('should stop propagation on a day key down event', () => {
      const e = { stopPropagation: jest.fn() };
      const x = '';
      instance.handleDayKeyDown(x, x, e);
      expect(instance.handleKeyDown).toHaveBeenCalledWith(e);
      expect(e.stopPropagation).toHaveBeenCalledTimes(1);
    });

  });

  describe('cancelDatePicker', () => {

    it('should sync props and close date picker', () => {
      wrapper.setState({ showDatePicker: true });
      instance.cancelDatePicker();
      expect(instance.syncTimeToState).toHaveBeenCalledWith(instance.props);
      expect(wrapper.state('showDatePicker')).toEqual(false);
    });

  });

  describe('showDatePicker', () => {

    it('should set date picker state', () => {
      wrapper.setState({ showDatePicker: false });
      instance.showDatePicker();
      expect(wrapper.state('showDatePicker')).toEqual(true);
    });

  });

  describe('handleDayClick', () => {

    it('should handle a day click while selecting', () => {
      const mockSelected = {};
      wrapper.setState({ selecting: true, selected: mockSelected, beforeSelected: null });
      instance.handleDayClick();
      expect(dateHelpers.getStartOfDay).not.toHaveBeenCalled();
      expect(wrapper.state('selected')).toBe(mockSelected);
      expect(wrapper.state('beforeSelected')).toBe(mockSelected);
      expect(wrapper.state('selecting')).toEqual(false);
    });

    it('should handle a day click when not selecting', () => {
      const mockSelected = {};
      const mockClicked = {};

      wrapper.setState({ selecting: false, selected: mockSelected, beforeSelected: null });
      const mockNewSelected = { from: 'start-of-day', to: 'end-of-day' };

      instance.handleDayClick(mockClicked);

      expect(dateHelpers.getStartOfDay).toHaveBeenCalledWith(mockClicked);
      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(mockClicked, { preventFuture: true });
      expect(wrapper.state('selected')).toEqual(mockNewSelected);
      expect(wrapper.state('beforeSelected')).toEqual(mockNewSelected);
      expect(wrapper.state('selecting')).toEqual(true);
    });

  });

  describe('handleDayHover', () => {

    it('should handle a day hover while selecting', () => {
      const mockOrderedRange = { from: 'from', to: 'to' };
      const mockHovered = {};
      instance.getOrderedRange = jest.fn(() => mockOrderedRange);
      wrapper.setState({ selecting: true });

      instance.handleDayHover(mockHovered);

      expect(instance.getOrderedRange).toHaveBeenCalledWith(mockHovered);
      expect(wrapper.state('selected')).toEqual(mockOrderedRange);
    });

    it('should handle a day hover when not selecting', () => {
      const mockHovered = {};
      const mockSelected = {};
      instance.getOrderedRange = jest.fn();
      wrapper.setState({ selecting: false, selected: mockSelected });

      instance.handleDayHover(mockHovered);

      expect(instance.getOrderedRange).not.toHaveBeenCalled();
      expect(wrapper.state('selected')).toEqual(mockSelected);
    });

  });

  describe('getOrderedRange', () => {

    it('should return correct range when new date is between from and to', () => {
      const from = new Date('2018-01-01');
      const newDate = new Date('2018-01-02');
      const to = new Date('2018-01-03');

      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(newDate, { preventFuture: true });
      expect(dateHelpers.getStartOfDay).not.toHaveBeenCalled();
      expect(range).toEqual({ from, to: 'end-of-day' });
    });

    it('should return correct range when new date is before from and to', () => {
      const newDate = new Date('2018-01-01');
      const from = new Date('2018-01-02');
      const to = new Date('2018-01-03');

      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).not.toHaveBeenCalled();
      expect(dateHelpers.getStartOfDay).toHaveBeenCalledWith(newDate);
      expect(range).toEqual({ from: 'start-of-day', to });
    });

    it('should return correct range when new date is after from and to', () => {
      const from = new Date('2018-01-01');
      const to = new Date('2018-01-02');
      const newDate = new Date('2018-01-03');

      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(newDate, { preventFuture: true });
      expect(dateHelpers.getStartOfDay).not.toHaveBeenCalled();
      expect(range).toEqual({ from, to: 'end-of-day' });
    });

    it('should return correct range when from and newDate are the same', () => {
      const from = new Date('2018-01-01');
      const newDate = new Date('2018-01-01');
      const to = new Date('2018-01-03');

      wrapper.setState({ beforeSelected: { from, to }});
      const range = instance.getOrderedRange(newDate);

      expect(dateHelpers.getEndOfDay).toHaveBeenCalledWith(newDate, { preventFuture: true });
      expect(dateHelpers.getStartOfDay).not.toHaveBeenCalled();
      expect(range).toEqual({ from, to: 'end-of-day' });
    });

  });

  describe('handleSelectRange', () => {

    let e;

    beforeEach(() => {
      e = {
        currentTarget: {}
      };
    });

    it('should open date picker for custom value', () => {
      e.currentTarget.value = 'custom';
      wrapper.setState({ showDatePicker: false });
      instance.handleSelectRange(e);
      expect(wrapper.state('showDatePicker')).toEqual(true);
      expect(props.onChange).not.toHaveBeenCalled();
    });

    it('should close date picker and refresh for non-custom value', () => {
      e.currentTarget.value = 'lasagna';
      wrapper.setState({ showDatePicker: true });
      instance.handleSelectRange(e);
      expect(wrapper.state('showDatePicker')).toEqual(false);
      expect(props.onChange).toHaveBeenCalledWith({ relativeRange: 'lasagna' });
    });

  });

  describe('handleFormDates', () => {

    it('should set from and to state and call the passed in callback', () => {
      wrapper.setState({ selected: false });
      const mockCallback = jest.fn();
      const mockSelected = { from: 'from', to: 'to' };
      instance.handleFormDates(mockSelected, mockCallback);
      expect(wrapper.state('selected')).toEqual(mockSelected);
      expect(mockCallback).toHaveBeenCalledTimes(1);
    });

  });

  describe('handleSubmit', () => {

    it('should reset some state and refresh', () => {
      const mockSelected = { a: 1, b: 2, c: 3 };
      wrapper.setState({ showDatePicker: true, selecting: true, selected: mockSelected });
      instance.handleSubmit();
      expect(wrapper.state('showDatePicker')).toEqual(false);
      expect(wrapper.state('selecting')).toEqual(false);
      expect(props.onChange).toHaveBeenCalledWith({ ...mockSelected, relativeRange: 'custom' });
    });

  });

});
