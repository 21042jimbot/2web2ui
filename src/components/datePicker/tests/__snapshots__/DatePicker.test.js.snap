// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: DatePicker should render ok by default 1`] = `
<Popover
  bottom={true}
  onOutsideClick={[Function]}
  open={false}
  right={true}
  trigger={
    <TextField
      connectLeft={
        <Select
          disabled={false}
          onChange={[Function]}
          options={
            Array [
              1,
              2,
              3,
            ]
          }
          placeholderValue=""
          required={false}
          value="day"
        />
      }
      labelHidden={true}
      onClick={[Function]}
      readOnly={true}
      required={false}
      resize="both"
      type="text"
      value="undefined - undefined"
    />
  }
  wrapper="div"
>
  <Datepicker
    disabledDays={
      Object {
        "after": 2017-01-15T12:00:00.000Z,
      }
    }
    enableOutsideDays={false}
    fixedWeeks={true}
    numberOfMonths={2}
    onDayClick={[Function]}
    onDayFocus={[Function]}
    onDayKeyDown={[Function]}
    onDayMouseEnter={[Function]}
    onKeyDown={[Function]}
    selectedDays={
      Object {
        "from": 2017-01-15T11:00:00.000Z,
        "to": 2017-01-15T12:00:00.000Z,
      }
    }
    toMonth={2017-01-15T12:00:00.000Z}
  />
  <ManualEntryForm
    from={2017-01-15T11:00:00.000Z}
    onEnter={[Function]}
    selectDates={[Function]}
    to={2017-01-15T12:00:00.000Z}
  />
  <Button
    onClick={[Function]}
    primary={true}
    size="default"
  >
    Apply
  </Button>
  <Button
    onClick={[Function]}
    size="default"
  >
    Cancel
  </Button>
</Popover>
`;
